design/baremetalhost-v1alpha2.md

```diff
@@ -0,0 +1,223 @@
+
+### Non-Goals
+
+- Describe all possible future API changes.
+- Describe a stable `v1` API.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to update the version of metal3 in
+my existing cluster without losing any data or features.
+
+## Design Details
+
+### Choosing a New Type for the Clock Speed Field
+
+The advice from the `controller-tools` maintainers is to use
+`resource.Quantity` or `int32` for floating point values.
+
+Our problematic field holds the clock speed of a CPU in fractional
+megahertz, so in order to use an integer we would need to represent
+the speed in terms of kilohertz.
+
+The `Quantity` type adds extra characters representing the unit
+conversion, so that a `1 GHz` value would be expressed as `1Gi`, which
+would be a bit ugly for someone trying to consume the value without
+access to the `go` library for parsing the string.
+
+The CPU speed is simply being reported, and not used by the
+`baremetal-operator` code. We could, therefore, use a `string` field
+without losing any functionality. The value we get from `gophercloud`
+is already a string.
+
+The `hardware-classification-controller` will need to convert the
+`string` to a `float64` before , but it would need to perform a similar
+conversion for `resource.Quantity`.
+
+### Converting Existing Data
+
+It is possible to [use a web
+hook](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#configure-customresourcedefinition-to-use-conversion-webhooks)
+to convert resources from their storage format to a specific API
+format. We do not want users to continue to consume the `v1alpha1`
+format, however, because we want to be able to drop support
+quickly. Therefore, we should not write a conversion web hook.
+
+Instead, we should add startup code to the baremetal-operator to read
+all host resources in their current format and write them (with type
+conversion but no other changes) to the new storage format. See option
+2
+[here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version)
+for more discussion.
+
+### Dropping Hardware Profile Field
+
+Now that we have support for more verbose root device hints and we
+have the `hardware-classification-controller`, we no longer need the
+hardware profile fields. We can remove them as part of the
+backwards-incompatible changes as we move to the new API version.
+
+### Risks and Mitigations
+
+We do not know all of the ways that consumers of the API are using the
+CPU speed values currently being reported. Given that most kubernetes
+tools appear to assume that floating point types are not to be used in
+APIs, it is likely that most consumers are either ignoring the value
+or using it in simple ways (displaying it, logging it, etc.) where
+loss of precision from the type conversion to a float is not
+significant.
+
```

Comment:
One possible source of risk is that the CAPM3/CAPBM needs to be updated in lockstep because we're dropping support for the old version immediately.

Reply:
That's a good point. There are "release" steps in the work items list. I should include "since to providers" as part of that.